builtin/stash--helper.c

```diff
@@ -0,0 +1,75 @@
+#include "../cache.h"
+#include "../cache-tree.h"
+#include "../diff.h"
+#include "../diffcore.h"
+#include "../gettext.h"
+#include "../git-compat-util.h"
+#include "../revision.h"
+
+static const char builtin_stash__helper_usage[] = {
+	"Usage: git stash--helper --non-patch <i_tree>"
+};
+
+static void diff_callback(struct diff_queue_struct *q,
+					 struct diff_options *options, void *data)
+{
+	int i;
+    fprintf(stderr,"diff_callback\n");
+	for (i = 0; i < q->nr; i++)
+    {
+		struct diff_filepair *p;
+		struct stat st;
+
+        fprintf(stderr,"diff_callback: %d\n", i);
+		p = q->queue[i];
+		add_to_index(&the_index, p->two->path, &st, 0);
+	}
+}
+
+int stash_non_patch(const char *i_tree,	const char *prefix)
+{
+	int result;
+	struct rev_info rev;
+	const char *wt_prefix = NULL;
+	unsigned char sha1[20];
+	const char *me = "git stash--helper";
+
+    fprintf(stderr,"stash_non_patch\n");
+	read_index(&the_index);
+    fprintf(stderr,"read the index\n");
+	init_revisions(&rev, NULL);
+    fprintf(stderr,"initialised rev\n");
+	setup_revisions(0, NULL, &rev, NULL);
+    fprintf(stderr,"setup rev\n");
+	rev.diffopt.output_format |= DIFF_FORMAT_CALLBACK;
+	rev.diffopt.format_callback = diff_callback;
+	rev.diffopt.format_callback_data=&result;
+	//diff_setup_done(&rev.diffopt);
+    fprintf(stderr,"prepared diffopts\n");
```

Comment:
Okay, I get the error now. With the `GIT_TRACE` output, I identified the failing call to be `git stash--helper --non-patch bb8058573351acfa9770d8e156931a8a44bf1a9e`. So I just re-ran that one in `gdb` and ended up with this:

```
$ (cd trash\ directory.t3903-stash/ && gdb -args ../../git 'stash--helper' '--non-patch' 'bb8058573351acfa9770d8e156931a8a44bf1a9e')
GNU gdb (Ubuntu 7.10-1ubuntu2) 7.10
[...]
(gdb) r
Starting program: /.../git/git stash--helper --non-patch bb8058573351acfa9770d8e156931a8a44bf1a9e
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
stash_non_patch
read the index
initialised rev
setup rev
prepared diffopts

Program received signal SIGSEGV, Segmentation fault.
0x00000000004d6839 in run_diff_index (revs=0x7fffffffd660, cached=0)
    at diff-lib.c:496
496		if (diff_cache(revs, ent->item->oid.hash, ent->name, cached))
(gdb) p ent
$1 = (struct object_array_entry *) 0x0
```

So it seems that we end up trying to dereference a `NULL` pointer...